torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py

```diff
@@ -0,0 +1,344 @@
+        raise RuntimeError(
+            "Overlapping DDP with ZeRO using this approach currently requires "
+            "NCCL backend to avoid hangs"
+        )
+
+    def hook_with_zero_fn(
+        state: Any,
+        bucket: dist.GradBucket,
+    ) -> torch.futures.Future:
+        r"""
+        Returns a :class:`Future` that gives a gradient bucket tensor and
+        performs the equivalent of a :class:`ZeroRedundancyOptimizer`
+        :meth:`step` if ``bucket`` is the last gradient bucket.
+
+        The function performs additional computation on the iteration that
+        the :class:`DistributedDataParallel` buckets are rebuilt to collect
+        information used to implement the modified hook.
+
+        Arguments:
+            state (Any): any state for the hook.
+            bucket (dist.GradBucket): the :class:`DistributedDataParallel`
+                gradient bucket.
+        """
+        fut = hook(state, bucket)
+        overlap_info = zero._overlap_info
+        bucket_index = bucket.get_index()
+
+        # Proceed as normal until the DDP buckets have been rebuilt
+        if not ddp._has_rebuilt_buckets:
+            assert overlap_info.status == _OverlapStatus.UNINITIALIZED
+            return fut
+
+        if overlap_info.status == _OverlapStatus.UNINITIALIZED:
+            overlap_info.status = _OverlapStatus.DDP_HAS_REBUILT_BUCKETS
+
+        rank = zero.global_rank
+        rank_to_update = zero._ddp_bucket_index_to_rank(bucket_index)
+
+        # Once DDP buckets have been rebuilt but ZeRO has not been
+        # properly initialized yet, collect the information needed
+        if overlap_info.status == _OverlapStatus.DDP_HAS_REBUILT_BUCKETS:
+            bucket_params = bucket.get_model_params_for_bucket()
+            assert len(bucket_params) > 0, "Empty bucket"
+            params_per_rank = overlap_info.params_per_rank
+            params_per_bucket = overlap_info.params_per_bucket
+            if rank_to_update == rank:
+                overlap_info.offsets[bucket_index] = len(params_per_rank[rank_to_update])
+            params_per_rank[rank_to_update].extend(bucket_params)
+            params_per_bucket.append(bucket_params)
+            return fut
+
+        assert overlap_info.status == _OverlapStatus.INITIALIZED
+
+        # Save the bucket reference and all-reduce future for the final bucket
+        if rank_to_update == rank:
+            overlap_info.bucket_index_to_bucket[bucket_index] = bucket
+            overlap_info.bucket_index_to_future[bucket_index] = fut
+
+        num_buckets = len(overlap_info.params_per_bucket)
+        # NOTE: The implementation from this point forward assumes that the
+        # buckets are indexed incrementally starting from 0 in the order of
+        # their autograd hooks firing
+        is_last_bucket = bucket_index == num_buckets - 1
+        if not is_last_bucket:
+            return fut
+
+        # Perform partial optimizer step on all buckets after the final
```

Comment:
Let's add a short comment to explain why this cannot be chained as a callback to the returned future.

Reply:
Added:
```
# NOTE: This should not be chained as a callback to the last bucket's
# all-reduce future since that would add synchronization that delays
# all optimizer computation to wait for that last all-reduce
```